EZEE-2880: Custom workflow actions (#952)
ezsystems/developer-documentation: docs/guide/extending/extending_workflow.md

# Extending Workflow

## Adding custom actions

[Built-in actions in the Editorial Workflow](../workflow.md#publishing-content-with-workflow)

> enable you to automatically publish a Content item or to sent a notification to reviewers.

Reviewer: ```suggestion
enable you to automatically publish a Content item or to send a notification to reviewers.
```

You can also create custom actions that will be called when content reaches a specific stage
or goes through a transition in a workflow.

``` yaml
custom_workflow:
    # ...
    transitions:
        # ...
        to_proofread:
            from: draft
            to: proofread
            label: To proofreading
            actions:
                proofread_transition_action:
                    data:
                        message: "The article has gone to proofreading"
```

The configuration indicates the name of the custom action (`proofread_transition_action`) that will call the action.

> `data` contains additional data that can be passed to the action. In this case it a message to display.

Reviewer: ```suggestion
`data` contains additional data that can be passed to the action. In this case, a message to display.
```

> To define what the action does, create an Event Listener, in this case `src/EventListener/ProofreadTransitionListener.php`:

Reviewer: ```suggestion
To define what the action does, create an Event Listener `src/EventListener/ProofreadTransitionListener.php`:
```

use Symfony\Component\Workflow\Event\TransitionEvent;
use EzSystems\EzPlatformAdminUi\Notification\TranslatableNotificationHandlerInterface as NotificationInterface;

class ProofreadTransitionListener extends AbstractTransitionWorkflowActionListener
{
    private $notificationHandler;

    public function __construct(NotificationInterface $notificationHandler)
    {
        $this->notificationHandler = $notificationHandler;
    }

    public function getIdentifier(): string
    {
        return 'proofread_transition_action';
    }

    public function onWorkflowEvent(TransitionEvent $event): void
    {
        $metadata = $this->getActionMetadata($event->getWorkflow(), $event->getTransition());
        $message = $metadata['data']['message'];

        $this->notificationHandler->info(
            $message,
            [],
            'domain'
        );

        $this->setResult($event, true);
    }
}
```

This Listener displays a notification bar at the bottom of the page when a Content item goes through the `to_proofread` transition.

The content of the notification is the message configured in `actions.proofread_transition_action.data`.

> To get it, you access the metadata for this transition through `getActionMetadata()` (line 27).

Reviewer: ```suggestion
To get it, you need to access the metadata for this transition through `getActionMetadata()` (line 27).
```
OR
```suggestion
To get it, access the metadata for this transition through `getActionMetadata()` (line 27).
```

use Symfony\Component\Workflow\Event\TransitionEvent;
use EzSystems\EzPlatformAdminUi\Notification\TranslatableNotificationHandlerInterface as NotificationInterface;

class DoneTransitionListener extends AbstractTransitionWorkflowActionListener
{
    private $notificationHandler;

    public function __construct(NotificationInterface $notificationHandler)
    {
        $this->notificationHandler = $notificationHandler;
    }

    public function getIdentifier(): string
    {
        return 'done_transition_action';
    }

    public function onWorkflowEvent(TransitionEvent $event): void
    {
        $context = $event->getContext();
        $message = $context['message'];

        $this->notificationHandler->info(
            $message,
            [],
            'domain'
        );
    }
}
```

This listener also displays a notification, but in this case its content is taken from the message
that the user types when choosing the `Done` transition.

The message is contained in the context of the action.

`$event->getContext()` gives you access to the context.
The context contains:

- `$workflowId` - the ID of the current workflow

> - `$message` - content of the message input by the user when sending the Content item through the transitions

Reviewer: ```suggestion
- `$message` - content of the user's message when sending the Content item through the transitions
```

``` php hl_lines="27 36"
<?php

declare(strict_types=1);

namespace App\EventListener;

use EzSystems\EzPlatformWorkflow\Event\Action\AbstractTransitionWorkflowActionListener;
use Symfony\Component\Workflow\Event\TransitionEvent;
use EzSystems\EzPlatformAdminUi\Notification\TranslatableNotificationHandlerInterface as NotificationInterface;

class ProofreadTransitionListener extends AbstractTransitionWorkflowActionListener
{
    private $notificationHandler;

    public function __construct(NotificationInterface $notificationHandler)
    {
        $this->notificationHandler = $notificationHandler;
    }

    public function getIdentifier(): string
    {
        return 'proofread_transition_action';
    }

    public function onWorkflowEvent(TransitionEvent $event): void
    {
        $metadata = $this->getActionMetadata($event->getWorkflow(), $event->getTransition());
        $message = $metadata['data']['message'];

        $this->notificationHandler->info(
            $message,
            [],
            'domain'
        );

        $this->setResult($event, true);

> 

Reviewer 2: Unnecessary empty line
